evm_arithmetization/src/cpu/kernel/asm/smt/delete.asm

```diff
@@ -0,0 +1,251 @@
+    %mload_trie_data
+    //stack: child_ptr, level, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    SWAP1 %increment SWAP1
+    //stack: child_ptr, level+1, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %jump(smt_delete_with_keys)
+
+// Update the internal node, possibly deleting it, or returning a leaf node.
+internal_update:
+    // Update the child first.
+    //stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP2 PUSH 1 SUB
+    //stack: 1-bit, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP4 ADD
+    //stack: sibling_ptr_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %mload_trie_data DUP1 %mload_trie_data
+    //stack: sibling_node_type, sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %eq_const(@SMT_NODE_HASH) %jumpi(sibling_is_hash)
+    %eq_const(@SMT_NODE_LEAF) %jumpi(sibling_is_leaf)
+sibling_is_internal:
+    //stack: sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+insert_child:
+    //stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %stack (deleted_child_ptr, bit, node_payload_ptr) -> (node_payload_ptr, bit, deleted_child_ptr, node_payload_ptr)
+    ADD %mstore_trie_data
+    // stack: node_payload_ptr, level, ks, retdest
+    %decrement
+    %stack (node_ptr, level, k0, k1, k2, k3, retdest) -> (retdest, node_ptr)
+    JUMP
+
+sibling_is_hash:
+    // stack: sibling_node_type, sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    //stack: sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %increment %mload_trie_data
+    // stack: hash, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jumpi(insert_child) // Sibling is non-empty hash node.
+sibling_is_empty:
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %mload_trie_data
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %eq_const(@SMT_NODE_HASH) %jumpi(sibling_is_empty_child_is_hash)
+    DUP1 %eq_const(@SMT_NODE_LEAF) %jumpi(sibling_is_empty_child_is_leaf)
+sibling_is_empty_child_is_internal:
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jump(insert_child)
+
+sibling_is_empty_child_is_hash:
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %increment %mload_trie_data
+    // stack: hash, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jumpi(insert_child)
```

Comment:
In that case, shouldn't we replace the whole internal node with the newly created hash node?

Reply:
Hm I don't think so, the hash node can hold an internal node.